doc/source/docs/user_guide/spatial_indexing.rst

```diff
@@ -0,0 +1,250 @@
+
+
+You can use any of the predicates available in :attr:`~sindex.SpatialIndex.valid_query_predicates`:
+
+.. ipython:: python
+
+    nyc.sindex.valid_query_predicates
+
+Array query
+~~~~~~~~~~~
+
+Checking a single geometry against the tree is nice but not that
+efficient if you are interested in many-to-many relationships. The
+:meth:`~sindex.SpatialIndex.query` method allows passing any 1-D array of geometries to be
+checked against the tree. If you do so, the output structure is slightly
+different:
+
+.. ipython:: python
+
+    bbox_array_query = nyc.sindex.query(points)
+    bbox_array_query
+
+
+By default, the method returns a 2-D array of indices where the query
+found a hit where the subarrays correspond to the indices of the input
+geometries and indices of the tree geometries associated with each. In
+the example above, the 0-th geometry in the ``points`` GeoSeries
+intersects the bounding box of the geometry at the position 1 from the
+``nyc`` GeoDataFrame, while the geometry 1 in the ``points`` matches
+geometry 16 in the ``nyc``. You may notice that these are the same
+indices as you’ve seen above.
+
+The other option is to return a boolean array with shape
+``(len(tree), n)`` with boolean values marking whether the bounding box
+of a geometry in the tree intersects a bounding box of a given geometry.
+This can be either a dense numpy array, or a sparse scipy array. Keep in
+mind that the output will be, in most cases, mostly filled with
+``False`` and the array can become really large, so it is recommended to
+use the sparse format, if possible.
+
+You can specify each using the ``output_format`` keyword:
+
+.. ipython:: python
+
+    bbox_array_query_dense = nyc.sindex.query(points, output_format="dense")
+    bbox_array_query_dense
+
+
+The dense array above has rows aligned with the rows of ``nyc`` and
+columns aligned with the rows of ``points`` and indicates all pairs
+where a *hit* was found.
+
+The same array can be represented as a :func:`scipy.sparse.coo_array`:
+
+.. ipython:: python
+
+    bbox_array_query_sparse = nyc.sindex.query(points, output_format="sparse")
+    bbox_array_query_sparse
+
+For example, to find the number of neighboring geometries for each subborough, you can
+use the spatial index to compare all geometries against each other. Since you are using
+``nyc`` on both sides of the query here, the resulting array is square-shaped with
+diagonal filled with ``True``.
+
+.. ipython:: python
+
+    neighbors = nyc.sindex.query(nyc.geometry, predicate="intersects", output_format="dense")
+    neighbors
+
+
+Getting the sum along one axis can then give you the answer. Note that
+since a geometry always intersects itself, you need to subtract one.
+
+.. ipython:: python
+
+    n_neighbors = neighbors.sum(axis=1) - 1
+    n_neighbors
+
+The result is a numpy array you can directly plot on a map.
+
+.. ipython:: python
+
+    nyc.plot(n_neighbors, legend=True);
```

Comment:
is this missing a savefig? I assume the intent is to show this in the docs, not just the code?